### PR TITLE
Adjust initialization in maximum and minimum

### LINKED ISSUE #27836: Performance degradation when reducing SparseMatrixCSC
I've found an extreme performance degradation (~30x slower) on Julia 0.7.0-beta vs. on Julia 0.6.2 when reducing a large `SparseMatrixCSC` matrix using `maximum`. I couldn't find differences in the data, so I think this is a problem in `SparseArrays`.

Julia 0.6.2
```
julia> size(counts)
(27998, 1306127)

julia> typeof(counts)
SparseMatrixCSC{Int32,Int64}

julia> @time maximum(counts, 2);
  9.171888 seconds (103.28 k allocations: 5.739 MiB, 0.06% gc time)

julia> @time sum(counts .> 0)
 10.630171 seconds (42.21 k allocations: 22.013 GiB, 0.08% gc time)
2624828308

julia> versioninfo()
Julia Version 0.6.2
Commit d386e40c17 (2017-12-13 18:08 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2637 v4 @ 3.50GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
```

Julia 0.7.0-beta:
```
julia> size(counts)
(27998, 1306127)

julia> typeof(counts)
SparseMatrixCSC{Int32,Int64}

julia> @time maximum(counts, dims=2);
271.014620 seconds (5.28 M allocations: 34.628 GiB, 0.09% gc time)

julia> @time sum(counts .> 0)
 10.284461 seconds (348.48 k allocations: 22.028 GiB, 0.69% gc time)
2624828308

julia> versioninfo()
Julia Version 0.7.0-beta.0
Commit f41b1ecaec (2018-06-24 01:32 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2637 v4 @ 3.50GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, broadwell)
```

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes #27836

### CODE REVIEW

File: base/reducedim.jl [hunk 1]
```diff
@@ -124,8 +105,33 @@ function _reducedim_init(f, op, fv, fop, A, region)
     return reducedim_initarray(A, region, z, Tr)
 end
 
-reducedim_init(f, op::typeof(max), A::AbstractArray{T}, region) where {T} = reducedim_initarray0(A, region, f, maximum)
-reducedim_init(f, op::typeof(min), A::AbstractArray{T}, region) where {T} = reducedim_initarray0(A, region, f, minimum)
+# initialization when computing minima and maxima requires a little care
+for (f1, f2, initval) in ((min, max, Inf), (max, min, -Inf))
+    function reducedim_init(f, op::typeof(f1), A::AbstractArray, region)
```
Comment (mbauman) [inline base/reducedim.jl]: This needs an `@eval` with interpolation — I don't think this will work as it stands.

File: base/reducedim.jl [hunk 2]
```diff
@@ -124,8 +105,33 @@ function _reducedim_init(f, op, fv, fop, A, region)
...
+        # First compute the reduce indices. This will throw an ArgumentError
+        # if any region is invalid
+        ri = Base.reduced_indices(A, region)
```
Comment (mbauman) [inline base/reducedim.jl]: `Base.`

**Comment by mbauman** [comment]
This seems like a good incremental improvement, but boy does this need a bigger revamp here.  It's somewhat nice that only a very limited set of functions are supported right now, because when we go to support arbitrary reduction functions we'll need to change the requirements — we'll need to require two slices along the dimension.

**Comment by andreasnoack** [comment]
I'd hope that the arrays initialization code could be cleaned up more generally. It is quite convoluted. Regarding your comment about supporting more reducers then an alternative to using two slices might be that users who happen to have a new `op` are asked to provide a function that returns a `y` such that for all `x` then `op(x, y) == x`. We currently utilize that these are `zero`, `one`, and `identity` for `+`, `*` and `min` respectively so it might work more generally.

Back to the current PR. There is some funny business going on with `OffsetArrays`. In the test, there is this array
```julia
julia> A = OffsetArray(rand(4,4), (-3,5));
```
What seems wrong to me is how `reduced_indices` work here since
```julia
julia> axes(A)
(Base.Slice(-2:1), Base.Slice(6:9))

julia> Base.reduced_indices(A, 1:2)
(Base.Slice(1:1), Base.Slice(1:1))
```
Notice that the last slice is not included in `Slice(6:9)`. Is this expected?

Finally, while trying to call any collection function on a `Slice` it is also clear how much we still rely on `size` in Base.

**Comment by mbauman** [comment]
> Is this expected?

As is often the case with OffsetArrays, we need to choose which behavior is more important: what should the axis of a reduced dimension be?  Do we collapse to `1:1` or do we collapse to the first index in the reduced dimension?  We've chosen the former, which has the unfortunate side-effect that the `reduced_indices` are _not_ a slice into the original object as you're trying to use it.

**Comment by andreasnoack** [comment]
It was a bit more tricky to finish up than expected but I think it works now. I've changed the non-`OneTo` case to return `Slice`s with the first index instead of `OneTo(1)`. I think it makes more sense. If we stick with the current behavior when we'd need a different fix for the initialization.

File: base/reducedim.jl [hunk 3]
```diff
@@ -124,8 +118,33 @@ function _reducedim_init(f, op, fv, fop, A, region)
...
+
+        # Make a view of the first slice of the region
+        A1 = view(A, ri...)
+
+        if isempty(A1)
+            # If the slice is empty just return non-view version as the initial array
+            return copy(A1)
+        else
+            # otherwise use the min/max of the first slice as initial value
+            v0 = mapreduce(f, $f2, A1)
```
Comment (nalimilan) [inline base/reducedim.jl]: It doesn't seem to be correct to use the min/max of the first slice as the initial value for all slices. I'm probably missing something. (I wish I could have used something simpler than what I did for this at https://github.com/JuliaLang/julia/pull/28027).

BTW, typo below: "intial"

Comment (andreasnoack) [reply base/reducedim.jl]: I think it should be fine. If you are computing the maximum along a dimension then it should be fine to initialize with any value less than or equal to the first element in the slice. The minimum over the first slice will satisfy this, right?

Comment (nalimilan) [reply base/reducedim.jl]: OK, I misunderstood "first slice". I thought it referred to the first slice along which reduction is performed. Carry on.

File: base/reducedim.jl [hunk 4]
```diff
@@ -124,8 +118,33 @@ function _reducedim_init(f, op, fv, fop, A, region)
...
+
+            # but NaNs need to be avoided as intial values
+            v0 = v0 != v0 ? typeof(v0)($initval) : v0
```
Comment (nalimilan) [inline base/reducedim.jl]: AFAICT the `mapreduce` call above will give `NaN` whenever one of the values is `NaN`. So this code is going to use the fallback `initval` even if there are non-`NaN` values which could be used. Maybe it would be better to use a custom loop to go over the first slice and skip `NaN` values?

Actually, maybe something similar to [the `mapfirst!` method](https://github.com/JuliaLang/julia/pull/28027/files#diff-c4a56019a8dae547ca5556ad8df4d1f0R358) I had to write at https://github.com/JuliaLang/julia/pull/27845 to skip missing values is needed.

Anyway, a few tests for these corner cases would be good. ;-)

Comment (andreasnoack) [reply base/reducedim.jl]: This is basically a check for floats but tries to be a little more generic. Using the `initval` is fine for floats and only floats will probably ever be `NaN` so I don't think there is an issue here.

**Comment by andreasnoack** [comment]
@timholy Are you okay with the reductions-on-offset-arrays behavior here?

**Comment by timholy** [comment]
Yes, it's quite elegant! Sorry I didn't see this earlier, thanks for tackling it.

File: base/reducedim.jl [hunk 5]
```diff
@@ -3,28 +3,41 @@
 ## Functions to compute the reduced shape
 
 # for reductions that expand 0 dims to 1
+reduced_index(i::OneTo) = OneTo(1)
+reduced_index(i::Slice) = Slice(first(i):first(i))
```
Comment (mbauman) [inline base/reducedim.jl]: It looks like simply making this a `UnitRange` instead of a `Slice` will do the trick — `make -C test reduce reduced offsetarray` passes for me locally with that change.
